cdap-app-fabric/src/main/java/io/cdap/cdap/internal/sysapp/SystemAppEnableExecutor.java

```diff
@@ -57,48 +57,62 @@
 
   @Inject
   SystemAppEnableExecutor(ApplicationLifecycleService appLifecycleService,
-      ProgramLifecycleService programLifecycleService) {
+                          ProgramLifecycleService programLifecycleService) {
     this.appLifecycleService = appLifecycleService;
     this.programLifecycleService = programLifecycleService;
   }
 
   public void deployAppAndStartPrograms(Arguments arguments) {
-    try {
-      // TODO(CDAP-16243): Make deploy app and start programs idempotent to not have transactional issues.
-      ApplicationWithPrograms appDetail = retryableDeploySystemApp(arguments);
-      for (ProgramDescriptor program : appDetail.getPrograms()) {
+
+
+    // TODO(CDAP-16243): Make deploy app and start programs idempotent to not have transactional issues.
+    ApplicationWithPrograms appDetail = retryableDeploySystemApp(arguments);
+    if(appDetail == null) return;
+
+    // TODO(CDAP-16243): Fix logic to restart programs for app with difference version/config.
+    for (ProgramDescriptor program : appDetail.getPrograms()) {
+      try {
         startProgram(program.getProgramId());
+      } catch (Exception ex) {
+        LOG.error("Failed to start program {} for app {} with exception {}",
+          program.getProgramId(), arguments.getId(), ex);
       }
-    } catch (Exception ex) {
-      LOG.error("Failed to deploy app {} and start its program with exception {}.",
-          arguments.getId(), ex);
-
     }
   }
 
-  private ApplicationWithPrograms retryableDeploySystemApp(Arguments arguments) throws Exception {
-    return Retries.callWithRetries(() -> deploySystemApp(arguments),
-        RetryStrategies.fixDelay(6, TimeUnit.SECONDS));
+  private ApplicationWithPrograms retryableDeploySystemApp(Arguments arguments) {
+    try {
+      return Retries.callWithRetries(() -> {
+            try {
+              return deploySystemApp(arguments);
+            } catch (RetryableException ex) {
+              LOG.debug("Failed to deploy app {} with exception {}. Will retry", arguments.getId(), ex);
+              ex.printStackTrace();
+            }
+            return null;
```

Comment:
this behavior means it don't actually get retried. It will catch the exception then return. You should use the version that takes a Predicate, and log there.